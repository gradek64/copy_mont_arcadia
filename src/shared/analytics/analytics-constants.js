import namespacedKeyMirror from '../lib/namespaced-key-mirror'

export const ANALYTICS_PLATFORM_ID = 'monty'

export const ANALYTICS_ERROR = {
  REGISTRATION_FAILED: 'User registration failed',
  LOGIN_FAILED: 'User login failed',
  SHOPPING_BAG_PROMO_CODE: 'Error applying promo code in shopping bag.',
  PROCEED_TO_PAYMENT: 'Error proceeding to payment',
  CONFIRM_AND_PAY: 'Error paying order',
  GIFT_CARD_ERROR: 'Error applying gift card in checkout',
}

export const ANALYTICS_FORM_ID = {
  REGISTER: 'register',
  LOGIN: 'login',
}

export const ANALYTICS_ACTION = namespacedKeyMirror('ANALYTICS', {
  SEND_API_RESPONSE_EVENT: null,
  SEND_CLICK_EVENT: null,
  SEND_ERROR_MESSAGE: null,
  SEND_PRODUCT_CLICK_EVENT: null,
  SEND_DISPLAY_EVENT: null,
  SEND_DELIVERY_OPTION_CHANGE_EVENT: null,
  SEND_INPUT_VALIDATION_STATUS: null,
  SEND_PAYMENT_METHOD_SELECTION_EVENT: null,
  SEND_PAYMENT_METHOD_INTENTION_EVENT: null,
  SEND_PAYMENT_METHOD_PURCHASE_SUCCESS_EVENT: null,
  SEND_PAYMENT_METHOD_PURCHASE_FAILURE_EVENT: null,
  SEND_DELIVERY_METHOD_CHANGE_EVENT: null,
  SEND_FILTER_USED_EVENT: null,
  SEND_ORDER_COMPLETE_EVENT: null,
  SEND_PURCHASE_EVENT: null,
  SEND_PURCHASE_ERROR_EVENT: null,
})

export const GTM_EVENT = {
  BAG_DRAWER_DISPLAYED: 'bagDrawerDisplayed',
  GEO_IP_MODAL_DISPLAYED: 'geoIpModal',
  CHECKOUT_SHIPPING_REDIRECT_MODAL_DISPLAYED: 'checkoutShippingRedirectModal',
}

export const GTM_TRIGGER = {
  PRODUCT_ADDED_TO_BAG: 'add to bag',
  PRODUCT_VIEW_BAG: 'view bag',
  BAG_ICON_CLICKED: 'bag icon',
  LOGIN: 'login',
}

export const GTM_CATEGORY = {
  SHOPPING_BAG: 'shoppingBag',
  CHECKOUT: 'checkout',
  SIGN_IN: 'signIn',
  REGISTER: 'register',
  HOME: 'home',
  PLP: 'plp',
  PDP: 'pdp',
  SEARCH_SELECTED: 'searchSelected',
  BUNDLE: 'bundle',
  STORE_LOCATOR: 'store-locator',
  FIND_IN_STORE: 'find-in-store',
  MR_CMS: 'mrCms-pages',
  PRODUCT_QUICK_VIEW: 'product-quick-view',
  DELIVERY_PAYMENT: 'delivery-payment',
  DELIVERY_DETAILS: 'delivery-details',
  COLLECT_FROM_STORE: 'collect-from-store',
  WRITE_REVIEW: 'write-a-review',
  REGISTER_LOGIN: 'register-login',
  BAG_DRAWER: 'bagDrawer',
  NOT_FOUND: 'not-found',
}

export const GTM_ACTION = {
  PROMO_CODE_APPLIED: 'promoCodeApplied',
  CLICKED: 'clicked',
  SUBMIT: 'submit',
  IMAGE_OVERLAY: 'image overlay',
  GIFT_CARD: 'giftCardApplied',
  CONTINUE_SHOPPING: 'continue shopping',
  SEARCH_BAR: 'searchbar',
  SIZE_SELECTED: 'selected size',
  BAG_DRAW_EDIT: 'bagDrawEdit',
  IN_STORE_POSTCODE_GO: 'inStorePostcodeGo',
  IN_STORE_POSTCODE_RESULT: 'inStorePostcodeResult',
  PAYMENT_METHOD_SELECTION: 'paymentMethodSelection',
  PAYMENT_METHOD_INTENTION: 'paymentMethodIntention',
  PAYMENT_METHOD_PURCHASE_SUCCESS: 'paymentMethodPurchaseSuccess',
  PAYMENT_METHOD_PURCHASE_FAILURE: 'paymentMethodPurchaseFailure',
}

export const GTM_PAGE_TYPES = {
  PDP_PAGE_TYPE: 'pdp',
  PLP_PAGE_TYPE: 'plp',
  PDP_BUNDLE_PAGE_TYPE: 'bundle',
  HOMEPAGE_PAGE_TYPE: 'homepage',
}

export const GTM_LIST_TYPES = {
  PDP_RECENTLY_VIEWED: 'PDP Recently Viewed',
  PLP_RECENTLY_VIEWED: 'PLP Recently Viewed',
  PDP_RECOMMENDED_PRODUCTS: 'PDP Recommended Products',
  // This variable is the same as PDP_RECOMMENDED_PRODUCTS, but it was decided to make it more specific in case we'll have another carousel of products on PDP page
  // PDP_RECOMMENDED_PRODUCTS variable could later be either deleted or renamed
  PDP_WHY_NOT_TRY: 'PDP Why Not Try',
}

export const GTM_VALIDATION_STATES = {
  SUCCESS: 'success',
  FAILURE: 'failure',
}

export const GTM_LABEL = {
  CLOSE_BUTTON: 'close-button',
  PRODUCT_DETAILS: 'product details',
}
