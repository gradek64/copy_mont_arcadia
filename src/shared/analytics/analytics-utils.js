export const calculatePayloadSize = (payload) => {
  const str = JSON.stringify(payload)
  // Matches only the 10.. bytes that are non-initial characters in a multi-byte sequence.
  const m = encodeURIComponent(str).match(/%[89ABab]/g)
  return (str ? str.length : 0) + (m ? m.length : 0)
}

/**
 * Takes a boolean and stringify's value for consumption by
 * google analytics
 * @param {boolan} bool
 * @returns {string} 'True' Or 'False'
 */
export const boolToString = (bool) => (bool ? 'True' : 'False')
