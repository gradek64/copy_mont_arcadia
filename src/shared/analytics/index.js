export {
  sendAnalyticsApiResponseEvent,
  sendAnalyticsClickEvent,
  sendAnalyticsErrorMessage,
  sendAnalyticsProductClickEvent,
  sendAnalyticsDisplayEvent,
  sendAnalyticsDeliveryOptionChangeEvent,
  sendAnalyticsValidationState,
  sendAnalyticsPaymentMethodSelectionEvent,
  sendAnalyticsPaymentMethodIntentionEvent,
  sendAnalyticsPaymentMethodPurchaseSuccessEvent,
  sendAnalyticsPaymentMethodPurchaseFailureEvent,
  sendAnalyticsDeliveryMethodChangeEvent,
  sendAnalyticsOrderCompleteEvent,
  sendAnalyticsPurchaseEvent,
} from './analytics-actions'

export {
  ANALYTICS_ERROR,
  GTM_CATEGORY,
  GTM_ACTION,
  GTM_LIST_TYPES,
  GTM_EVENT,
  GTM_TRIGGER,
  GTM_VALIDATION_STATES,
  GTM_LABEL,
  GTM_PAGE_TYPES,
} from './analytics-constants'
