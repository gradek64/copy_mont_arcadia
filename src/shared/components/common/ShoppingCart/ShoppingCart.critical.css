/** @define ShoppingCart; weak */

.ShoppingCart {
  display: flex;
  height: 100%;
  margin: $shopping-cart-margin-tablet;
  overflow: visible;

  &-icon {
    background-image: $shopping-cart-icon-active;
    background-position: center;
    background-repeat: no-repeat;
    background-size: contain;
    display: flex;
    justify-content: center;
    height: $shopping-cart-icon-height-tablet;
    position: relative;
    vertical-align: middle;
    width: $shopping-cart-icon-width-tablet;
    align-items: $shopping-cart-align-items;
  }

  &.is-empty {
    & .ShoppingCart-icon {
      background-image: $shopping-cart-icon;
    }
  }

  &-itemsCount {
    background-color: $shopping-cart-count-bg-color;
    border-radius: 100%;
    color: $shopping-cart-count-color;
    display: $shopping-cart-count-display;
    font-size: $shopping-cart-count-font-size-tablet;
    left: $shopping-cart-count-left;
    line-height: $shopping-cart-count-line-height;
    min-width: $shopping-cart-count-min-width;
    position: $shopping-cart-count-position;
    top: $shopping-cart-count-top;
    vertical-align: $shopping-cart-count-vertical-align;
    margin-top: $shopping-cart-count-margin-top;
    align-self: $shopping-cart-count-align-self;
    margin-bottom: $shopping-cart-count-margin-bottom;
  }
}

@media $min-laptop {
  .ShoppingCart {

    &-icon {
      display: flex;
      justify-content: center;
    }
  }
}

@media $desktop {
  .ShoppingCart {
    margin: $shopping-cart-margin;

    &-icon {
      height: $shopping-cart-icon-height;
      width: $shopping-cart-icon-width;
      display: flex;
      justify-content: center;
    }

    &-itemsCount {
      font-size: $shopping-cart-count-font-size;
      top: $shopping-cart-count-top-desktop;
      line-height: $shopping-cart-count-line-height-desktop;
      vertical-align: $shopping-cart-count-vertical-align-desktop;
    }
  }
}
