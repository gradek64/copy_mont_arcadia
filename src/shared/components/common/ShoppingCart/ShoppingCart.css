/** @define ShoppingCart; weak */

.ShoppingCart {

  &-label {
    color: $font-color-base;
    display: none;
    font-size: $shopping-cart-label-font-size-tablet;
    margin: 0 5px 0 0;
    vertical-align: middle;
    align-self: center;
  }

}

@media $min-laptop {
  .ShoppingCart {
    &-label {
      display: inline;
    }
  }
}

@media $desktop {
  .ShoppingCart {

    &-label {
      font-size: $shopping-cart-label-font-size;
    }

  }
}

@media $min-laptop {
  .ShoppingCart {
    &-label {
      margin: $shopping-cart-label-margin;
    }
  }
}
