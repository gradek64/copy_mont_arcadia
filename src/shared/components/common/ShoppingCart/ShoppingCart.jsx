import classnames from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'
import { connect } from 'react-redux'
import {
  sendAnalyticsDisplayEvent,
  GTM_EVENT,
  GTM_TRIGGER,
} from '../../../analytics'
import { toggleMiniBag } from '../../../../shared/actions/common/shoppingBagActions'

const mapStateToProps = (state) => ({
  itemsCount: state.shoppingBag.totalItems,
  pageType: state.pageType,
})

const mapDispatchToProps = {
  onClick: toggleMiniBag,
  sendAnalyticsDisplayEvent,
}

const ShoppingCart = ({
  label,
  itemsCount,
  modifier,
  className,
  onClick,
  sendAnalyticsDisplayEvent,
  pageType,
}) => {
  const classNames = classnames('ShoppingCart', className, {
    [`ShoppingCart--${modifier}`]: modifier,
    'is-empty': !itemsCount,
  })

  return (
    <button
      className={classNames}
      onClick={() => {
        onClick()
        sendAnalyticsDisplayEvent(
          {
            bagDrawerTrigger: GTM_TRIGGER.BAG_ICON_CLICKED,
            openFrom: pageType,
          },
          GTM_EVENT.BAG_DRAWER_DISPLAYED
        )
      }}
    >
      {label && <span className="ShoppingCart-label">{label}</span>}
      <span className="ShoppingCart-icon">
        {itemsCount > 0 && (
          <span className="ShoppingCart-itemsCount">{itemsCount}</span>
        )}
      </span>
    </button>
  )
}

ShoppingCart.propTypes = {
  label: PropTypes.string,
  itemsCount: PropTypes.number,
  modifier: PropTypes.string,
  className: PropTypes.string,
  onClick: PropTypes.func.isRequired,
  sendAnalyticsDisplayEvent: PropTypes.func.isRequired,
  pageType: PropTypes.string,
}

ShoppingCart.defaultProps = {
  label: undefined,
  itemsCount: 0,
  modifier: undefined,
  className: '',
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ShoppingCart)

export { ShoppingCart as WrappedShoppingCart }
