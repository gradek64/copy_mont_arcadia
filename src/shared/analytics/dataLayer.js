import pageSchema from './schemas/page'
import userSessionSchema from './schemas/userSession'
import basketSchema from './schemas/basket'
import promosSchema from './schemas/promos'
import { productClickSchema } from './schemas/product'
import { ANALYTICS_PLATFORM_ID } from './analytics-constants'

let dataLayerModule = null

if (process.env.NODE_ENV === 'production') {
  dataLayerModule = require('@ag-digital/analytics-datalayer/prod')
} else {
  dataLayerModule = require('@ag-digital/analytics-datalayer/dev')
}

const dataLayer = dataLayerModule.default

const origCall = dataLayer.push
dataLayer.push = (data, schemaName, eventName) => {
  origCall({ ...data, platform: ANALYTICS_PLATFORM_ID }, schemaName, eventName)
}

// GTM-136 🐒 - So That undefined can be passed into the schema (So that it passes validation)
if (process.env.NODE_ENV !== 'production') {
  const origCall = dataLayer.push
  dataLayer.push = (data, schemaName, eventName) => {
    origCall(JSON.parse(JSON.stringify(data)), schemaName, eventName)
  }
}

// Mock adapter in case this is run server-side without a `window` object
export const attachable =
  typeof window === 'undefined' ? { push: () => {}, dataLayer: [] } : window

// This adapter pushes data to GTM in the way it expects
export const gtm = dataLayerModule.gtmAdapter(attachable)
export const extendArray = dataLayerModule.extendArray

if (process.browser) {
  // Schemas are based on the http://json-schema.org/ standard
  // To learn: https://spacetelescope.github.io/understanding-json-schema/
  dataLayer
    // Attaches the dataLayer to `window.analytics-datalayer` so it can be tested by an E2E framework
    // in the future
    .attachTo(attachable, 'analytics-datalayer')
    .addSchema('pageSchema', pageSchema)
    .addSchema('userSessionSchema', userSessionSchema)
    .addSchema('basketSchema', basketSchema)
    .addSchema('promosSchema', promosSchema)
    .addSchema('productClickSchema', productClickSchema)
    .addAdapter('gtm', gtm)
}

export default dataLayer
