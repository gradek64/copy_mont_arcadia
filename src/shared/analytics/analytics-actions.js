import { ANALYTICS_ACTION, GTM_ACTION } from './analytics-constants'
import { getCurrencyCode } from '../selectors/configSelectors'
import {
  getCheckoutOrderError,
  getCheckoutOrderId,
  getCheckoutOrderLines,
  getCheckoutUserId,
  getCheckoutPromoCodes,
  getCheckoutTotalOrderPrice,
  getCheckoutTotalOrderDiscount,
  getCheckoutProductRevenue,
  getCheckoutPaymentDetails,
  getCheckoutDeliveryAddress,
  getCheckoutDeliveryMethod,
  getCheckoutDeliveryPrice,
  getDDPPromotion,
  getCheckoutUserType,
  getIsRegisteredEmail,
} from '../selectors/checkoutSelectors'
import transformBasket from '../../client/analytics/transforms/basket'
import { round } from '../lib/numbers'
import { getUserAuthState } from '../selectors/userAuthSelectors'
import { getDDPUserAnalyticsProperties } from '../selectors/ddpSelectors'
import {
  getUserTrackingId,
  getHashedUserEmail,
} from '../selectors/common/accountSelectors'
import { boolToString } from './analytics-utils'

export const sendAnalyticsApiResponseEvent = (payload) => ({
  type: ANALYTICS_ACTION.SEND_API_RESPONSE_EVENT,
  payload,
})

export const sendAnalyticsClickEvent = (payload) => ({
  type: ANALYTICS_ACTION.SEND_CLICK_EVENT,
  payload,
})

export const sendAnalyticsErrorMessage = (errorMessage) => ({
  type: ANALYTICS_ACTION.SEND_ERROR_MESSAGE,
  errorMessage,
})

export const sendAnalyticsProductClickEvent = (payload) => ({
  type: ANALYTICS_ACTION.SEND_PRODUCT_CLICK_EVENT,
  payload,
})

export const sendAnalyticsDisplayEvent = (payload, eventName) => ({
  type: ANALYTICS_ACTION.SEND_DISPLAY_EVENT,
  payload,
  eventName,
})

export const sendAnalyticsDeliveryOptionChangeEvent = (
  deliveryLocationType
) => ({
  type: ANALYTICS_ACTION.SEND_DELIVERY_OPTION_CHANGE_EVENT,
  deliveryLocationType,
})

export const sendAnalyticsValidationState = ({ id, validationStatus }) => ({
  type: ANALYTICS_ACTION.SEND_INPUT_VALIDATION_STATUS,
  id,
  validationStatus,
})

export const sendAnalyticsPaymentMethodSelectionEvent = (payload) => ({
  type: ANALYTICS_ACTION.SEND_PAYMENT_METHOD_SELECTION_EVENT,
  eventName: GTM_ACTION.PAYMENT_METHOD_SELECTION,
  payload,
})

export const sendAnalyticsPaymentMethodIntentionEvent = (payload) => ({
  type: ANALYTICS_ACTION.SEND_PAYMENT_METHOD_INTENTION_EVENT,
  eventName: GTM_ACTION.PAYMENT_METHOD_INTENTION,
  payload,
})

export const sendAnalyticsPaymentMethodPurchaseSuccessEvent = (payload) => ({
  type: ANALYTICS_ACTION.SEND_PAYMENT_METHOD_PURCHASE_SUCCESS_EVENT,
  eventName: GTM_ACTION.PAYMENT_METHOD_PURCHASE_SUCCESS,
  payload,
})

export const sendAnalyticsPaymentMethodPurchaseFailureEvent = (payload) => ({
  type: ANALYTICS_ACTION.SEND_PAYMENT_METHOD_PURCHASE_FAILURE_EVENT,
  eventName: GTM_ACTION.PAYMENT_METHOD_PURCHASE_FAILURE,
  payload,
})

export const sendAnalyticsDeliveryMethodChangeEvent = (deliveryMethod) => ({
  type: ANALYTICS_ACTION.SEND_DELIVERY_METHOD_CHANGE_EVENT,
  deliveryMethod,
})

export const sendAnalyticsFilterUsedEvent = (payload) => ({
  type: ANALYTICS_ACTION.SEND_FILTER_USED_EVENT,
  payload,
})

export const sendAnalyticsOrderCompleteEvent = (payload) => ({
  type: ANALYTICS_ACTION.SEND_ORDER_COMPLETE_EVENT,
  payload,
})

export const sendAnalyticsPurchaseEvent = () => (dispatch, getState) => {
  const state = getState()

  // From config
  const currencyCode = getCurrencyCode(state)

  // From checkout
  const orderError = getCheckoutOrderError(state)

  // From orderCompleted
  const orderId = getCheckoutOrderId(state)

  if (orderError || !orderId) {
    return dispatch({
      type: ANALYTICS_ACTION.SEND_PURCHASE_ERROR_EVENT,
      payload: {
        orderError,
        orderId,
      },
    })
  }

  // From orderCompleted
  const loggedIn = boolToString(getUserTrackingId(state))
  const authState = getUserAuthState(state)
  const userId = getCheckoutUserId(state)
  const hashedEmailAddress = getHashedUserEmail(state)
  const orderLines = getCheckoutOrderLines(state)
  const promoCodes = getCheckoutPromoCodes(state)
  const totalOrderPrice = parseFloat(getCheckoutTotalOrderPrice(state))
  const productRevenue = parseFloat(getCheckoutProductRevenue(state))
  const paymentDetails = getCheckoutPaymentDetails(state)
  const deliveryAddress = getCheckoutDeliveryAddress(state)
  const deliveryMethod = getCheckoutDeliveryMethod(state)
  const deliveryPrice = parseFloat(getCheckoutDeliveryPrice(state))
  const totalOrdersDiscount = getCheckoutTotalOrderDiscount(state)
  const orderWithDDPPromotion = getDDPPromotion(state)
  const userType = getCheckoutUserType(state)
  const isRegisteredEmail = getIsRegisteredEmail(state)

  const twoDP = (value) =>
    typeof value === 'number' && !Number.isNaN(value)
      ? round(value).toFixed(2)
      : '0.00'

  const paymentType = paymentDetails.map((item) => item.paymentMethod).join(',')

  const calculateOrderDiscount = (
    totalOrdersDiscount,
    totalOrderPrice,
    deliveryPrice
  ) => {
    const trimToNumber = (string) =>
      string.replace(/^[^\d]+/, '').replace(/[^\d]+$/, '')

    const orderDiscount = parseFloat(trimToNumber(totalOrdersDiscount))

    if (Number.isNaN(orderDiscount) || orderDiscount === 0) {
      return 0
    }

    const productsPrice = totalOrderPrice - deliveryPrice
    return (orderDiscount / (productsPrice + orderDiscount)) * 100
  }

  const actionField = {
    id: `${orderId}`,
    revenue: twoDP(totalOrderPrice),
    productRevenue: twoDP(productRevenue),
    paymentType,
    orderDiscount: twoDP(
      calculateOrderDiscount(
        totalOrdersDiscount,
        totalOrderPrice,
        deliveryPrice
      )
    ),
    shippingCountry: deliveryAddress.country,
    shippingOption: deliveryMethod,
    shipping: twoDP(deliveryPrice),
    ddpOrder: {
      ddpPromotion: Object.keys(orderWithDDPPromotion).length > 0,
      value: orderWithDDPPromotion.value || 0,
    },
  }

  if (promoCodes.length) {
    // GA Measurement Protocol can only accommodate one coupon code
    // but this array can be transformed in GTM for consumption by
    // GA and any other analytics service that requires it.
    actionField.coupon = promoCodes
  }

  const products = orderLines.map(
    ({
      lineNo,
      skuId,
      name,
      unitPrice,
      colour,
      quantity,
      category,
      size,
      brand,
      reviewRating,
      ecmcCategory,
      department,
    }) => ({
      id: lineNo,
      productId: skuId,
      name: `(${lineNo}) ${name}`,
      price: unitPrice,
      unitNowPrice: unitPrice,
      colour,
      quantity: `${quantity}`,
      category,
      size,
      brand,
      reviewRating,
      ecmcCategory,
      department,
    })
  )

  const payload = {
    user: {
      ...getDDPUserAnalyticsProperties(state),
      loggedIn,
      authState,
      id: userId,
      userType,
      isRegisteredEmail,
    },
    ecommerce: {
      currencyCode,
      purchase: {
        actionField,
        products,
      },
    },

    // Exponea needs an empty basket to clear its state.
    fullBasket: transformBasket(state),

    // Prevent merging of the payload properties (effectively flush)
    // that are siblings to this data layer command.
    _clear: true,
  }

  if (hashedEmailAddress) {
    payload.user.hashedEmailAddress = hashedEmailAddress
  }

  return dispatch({
    type: ANALYTICS_ACTION.SEND_PURCHASE_EVENT,
    payload,
  })
}
