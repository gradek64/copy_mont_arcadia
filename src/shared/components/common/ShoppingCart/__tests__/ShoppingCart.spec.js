import testComponentHelper from 'test/unit/helpers/test-component'
import { GTM_TRIGGER, GTM_EVENT } from '../../../../analytics'

import { WrappedShoppingCart } from '../ShoppingCart'

describe('<ShoppingCart />', () => {
  const requiredProps = {
    onClick: () => {},
  }
  const renderComponent = testComponentHelper(WrappedShoppingCart)

  describe('@renders', () => {
    it('should render in default state', () => {
      const { getTree } = renderComponent(requiredProps)
      expect(getTree()).toMatchSnapshot()
    })

    it('should display `label` if prop supplied', () => {
      const { wrapper } = renderComponent({
        ...requiredProps,
        label: 'Label',
      })
      expect(wrapper.find('.ShoppingCart-label')).toHaveLength(1)
      expect(wrapper.find('.ShoppingCart-label').text()).toBe('Label')
    })

    it('should display `itemsCount` if prop supplied', () => {
      const { wrapper } = renderComponent({
        ...requiredProps,
        itemsCount: 5,
      })
      expect(wrapper.find('.ShoppingCart-itemsCount').text()).toBe('5')
    })

    it('should add `is-empty` class name if itemsCount prop is `0`', () => {
      const { wrapper } = renderComponent({
        ...requiredProps,
        itemsCount: 0,
      })
      expect(wrapper.find('.ShoppingCart').hasClass('is-empty')).toBe(true)
    })

    it('should add `modifier` to class name if prop supplied', () => {
      const { wrapper } = renderComponent({
        ...requiredProps,
        modifier: 'modified',
      })
      expect(
        wrapper.find('.ShoppingCart').hasClass('ShoppingCart--modified')
      ).toBe(true)
    })
  })

  describe('@events', () => {
    it('should call `onClick` prop on click', () => {
      const props = {
        onClick: jest.fn(),
        sendAnalyticsDisplayEvent: jest.fn(),
        pageType: 'plp',
      }
      const { wrapper } = renderComponent(props)
      wrapper.prop('onClick')()
      expect(props.onClick).toHaveBeenCalled()

      expect(props.sendAnalyticsDisplayEvent).toHaveBeenCalledWith(
        {
          bagDrawerTrigger: GTM_TRIGGER.BAG_ICON_CLICKED,
          openFrom: 'plp',
        },
        GTM_EVENT.BAG_DRAWER_DISPLAYED
      )
    })
  })
})
