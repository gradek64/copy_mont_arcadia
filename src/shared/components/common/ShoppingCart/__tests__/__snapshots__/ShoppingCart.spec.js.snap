// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShoppingCart /> @renders should render in default state 1`] = `
<button
  className="ShoppingCart is-empty"
  onClick={[Function]}
>
  <span
    className="ShoppingCart-icon"
  />
</button>
`;
